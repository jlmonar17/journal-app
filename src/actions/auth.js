import Swal from "sweetalert2";
import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";

export const smartLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());

        firebase
            .auth()
            .signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch((e) => {
                Swal.fire("Error", e.message, "error");
            })
            .finally(() => {
                dispatch(finishLoading());
            });
    };
};

// Asynchronus action, after resolve, it will dispatch another SYNCHRONUS action.
// Asynchronus actions return a callback. in difference to SYNCHRONUS actions.
export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase
            .auth()
            .signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch((e) => console.log(e));
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase
            .auth()
            .createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name });

                dispatch(login(user.uid, user.displayName));
            })
            .catch((e) => {
                Swal.fire("Error", e.message, "error");
            });
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());
    };
};

export const logout = () => ({ type: types.logout });
