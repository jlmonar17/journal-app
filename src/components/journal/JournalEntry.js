export const JournalEntry = () => {
    return (
        <div className="journal__entry pointer">
            <div
                className="journal__entry-picture"
                style={{
                    backgroundSize: "cover",
                    backgroundImage:
                        "url(https://iso.500px.com/wp-content/uploads/2014/07/big-one.jpg)",
                }}
            ></div>

            <div className="journal__entry-body">
                <p className="journal__entry-title">Un nuevo dia</p>

                <p className="journal__entry-content">
                    Lorem ipsum dolor sit amet
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>Monday</span>
                <h4>22</h4>
            </div>
        </div>
    );
};
