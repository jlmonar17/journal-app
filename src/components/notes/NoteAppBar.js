import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploadingFile } from "../../actions/notes";

export const NoteAppBar = () => {
    const dispatch = useDispatch();
    const { active } = useSelector((state) => state.notes);

    const handleSaveNote = () => {
        const note = { ...active };
        delete note.url;

        dispatch(startSaveNote(note));
    };

    const handlePictureClick = () => {
        document.querySelector("#fileSelector").click();
    };

    const handleFileChange = (e) => {
        const file = e.target.files[0];

        dispatch(startUploadingFile(file));
    };

    return (
        <div className="notes__appbar">
            <span>23 de Julio 2021</span>

            <input
                id="fileSelector"
                type="file"
                style={{ display: "none" }}
                onChange={handleFileChange}
            />

            <div>
                <button className="btn" onClick={handlePictureClick}>
                    Picture
                </button>
                <button className="btn" onClick={handleSaveNote}>
                    Save
                </button>
            </div>
        </div>
    );
};
