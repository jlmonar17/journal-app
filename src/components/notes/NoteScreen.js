import { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNote } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NoteAppBar } from "./NoteAppBar";

export const NoteScreen = () => {
    const dispatch = useDispatch();

    const { active: note } = useSelector((state) => state.notes);
    const { formValues, handleInputChange, reset } = useForm(note);

    // useRef returns a ref object mutable, and the property .current will be initialized with argument
    // passed (initialValue). Object returned will keep persistent during all component lifecycle
    // useRef keeps the variable mutable and wont rerender component if value changes.
    const activeId = useRef(note.id);

    useEffect(() => {
        if (note.id !== activeId.current) {
            reset(note);

            // It will be executed only if active note changes.
            // We change .current property of mutable object, it won't cause component re render.
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch(activeNote(formValues.id, { ...formValues }));
    }, [formValues, dispatch]);

    const { title, body } = formValues;

    return (
        <div className="notes__main-content">
            <NoteAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    value={title}
                    onChange={handleInputChange}
                />

                <textarea
                    placeholder="What happened today?"
                    className="notes__textarea"
                    name="body"
                    value={body}
                    onChange={handleInputChange}
                ></textarea>

                {note.url && (
                    <div className="notes__image">
                        <img src={note.url} alt="note" />
                    </div>
                )}
            </div>
        </div>
    );
};
