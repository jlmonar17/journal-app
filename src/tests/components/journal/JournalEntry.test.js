import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { JournalEntry } from "../../../components/journal/JournalEntry";
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

// We set initial state desired for Redux state.
const initialState = {};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const note = {
    id: 1,
    title: "Hello",
    body: "World",
    date: 0,
    url: "https://test-image.com/image.jpg",
};

describe("Tests for <JournalEntry />", () => {
    const wrapper = mount(
        <Provider store={store}>
            <JournalEntry {...note} />
        </Provider>
    );

    test("Should render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("Should active note", () => {
        wrapper.find(".journal__entry").prop("onClick")();

        expect(store.dispatch).toHaveBeenCalledWith(activeNote(note.id, note));
    });
});
