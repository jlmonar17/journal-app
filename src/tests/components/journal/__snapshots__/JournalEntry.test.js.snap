// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <JournalEntry /> Should render correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://test-image.com/image.jpg)",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hello
    </p>
    <p
      className="journal__entry-content"
    >
      World
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
